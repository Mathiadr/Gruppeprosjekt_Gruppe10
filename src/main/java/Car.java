import java.util.ArrayList;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether or not the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    private String fuelType;        //TODO: IMPLEMENT
    private String transmission;    //TODO: IMPLEMENT

    private ArrayList<String> features;


    private static ArrayList<Car> carArrayList = new ArrayList<>();

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }


    public String getAvailableDate() {
        return availableDate;
    }

    public void setAvailableDate(String availableDate) {
        this.availableDate = availableDate;
    }

    public static ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public static void setCarArrayList(ArrayList<Car> carArrayList) {
        Car.carArrayList = carArrayList;
    }

    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
