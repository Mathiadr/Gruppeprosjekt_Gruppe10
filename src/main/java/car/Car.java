package car;

import java.util.ArrayList;
import java.util.Collection;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    //private String availableDate;   //TODO: IMPLEMENT
    private boolean isAvailable = true; //Whether the car is occupied (Not to be confused with available date)
                                        //TODO: IMPLEMENT
    //private String fuelType;        //TODO: IMPLEMENT
    //private String transmission;    //TODO: IMPLEMENT

    //private Collection<String> features; //TODO: IMPLEMENT

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }
}
