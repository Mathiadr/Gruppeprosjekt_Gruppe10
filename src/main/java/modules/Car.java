package modules;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Date;

public class Car {
    private String registrationNumber;
    private String owner;
    private String model;
    private String brand;                   //TODO: IMPLEMENT
    //private String fuelType;              //TODO: IMPLEMENT
    //private String transmission;          //TODO: IMPLEMENT
    //private Collection<String> features;  //TODO: IMPLEMENT
    private Listing listing;

    public static final String[] fuelTypesList = {"Electric", "Gasoline", "Diesel"};
    public static final String[] transmissionList = {"Manual", "Automatic"};
    public static final String[] featureList = {"GPS", "Air Conditioning"};



    public Car(){

    }

    public Car(String registrationNumber, String owner, String model) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        LocalDate startDateplaceholder = LocalDate.of(2022, 1, 1);
        LocalDate endDatePlaceholder = LocalDate.of(2022, 12, 31);
        this.listing = new Listing(startDateplaceholder, endDatePlaceholder, true, "");
    }
    public Car(String registrationNumber, String owner, String model, Listing listing) {
        this.registrationNumber = registrationNumber;
        this.owner = owner;
        this.model = model;
        this.listing = listing;
    }


    public String getRegistrationNumber() {
        return registrationNumber;
    }

    public void setRegistrationNumber(String registrationNumber) {
        this.registrationNumber = registrationNumber;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public Listing getListing() {
        return listing;
    }

    public void setListing(Listing listing) {
        this.listing = listing;
    }



    @Override
    public String toString(){
        return "\n\nregistrationNumber: " + registrationNumber + ",\n" +
                "Owner: " + owner + ",\n" +
                "Model: " + model + ",\n" +
                "Listing: " + listing;
    }
}
