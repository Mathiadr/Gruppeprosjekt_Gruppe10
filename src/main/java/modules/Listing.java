package modules;


import java.util.Date;

public class Listing {
    // Contains information of every bid, each of which belongs to one car.
    // TODO: EVERY CAR BELONGS TO ONE LISTING
    // TODO: ur mum
    private String publisherUsername;
    private Date startDate;
    private Date endDate;
    private Car car;
    private boolean isAvailable;
    private boolean published;
    private String description;

    public Listing(){}

    public Listing(String publisherUsername, Date startDate, Date endDate, Car car, boolean published, String description){
        this.publisherUsername = publisherUsername;
        this.startDate = startDate;
        this.endDate = endDate;
        this.car    = car;
        this.isAvailable = true;
        this.published = published;
        this.description = description;
    }


    public Date getStartDate() {
        return startDate;
    }

    public void setStartDate(Date startDate) {
        this.startDate = startDate;
    }

    public Date getEndDate() {
        return endDate;
    }

    public void setEndDate(Date endDate) {
        this.endDate = endDate;
    }

    public Car getCar() {
        return car;
    }

    public void setCar(Car car) {
        this.car = car;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public void setAvailable(boolean available) {
        isAvailable = available;
    }

    public boolean isPublished() {
        return published;
    }

    public void setPublished(boolean published) {
        this.published = published;
    }



}
