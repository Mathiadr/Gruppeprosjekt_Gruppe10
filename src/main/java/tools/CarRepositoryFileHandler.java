package tools;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import interfaces.FileHandler;
import modules.Car;


import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

public class CarRepositoryFileHandler implements FileHandler<Car> {

    @Override
    public ArrayList<Car> readFromFile(String filename) {
        ObjectMapper objectMapper = new ObjectMapper();
        File file = new File(filename);
        try {
            return objectMapper.readValue(file, new TypeReference<ArrayList<Car>>() {});
        } catch (IOException e) {
            System.err.println(e);
        }
        return null;
    }

    @Override
    public void writeArrayListToFile(ArrayList<Car> carArrayList, String filename) {
        ObjectMapper objectMapper = new ObjectMapper();
        File file = new File(filename);

        try {
            objectMapper.writerWithDefaultPrettyPrinter().writeValue(file, carArrayList);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}