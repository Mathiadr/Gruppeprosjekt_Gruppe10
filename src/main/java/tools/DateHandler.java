package tools;

import com.spun.util.velocity.ParserDateUtils;
import modules.Car;
import modules.Listing;

import javax.swing.*;
import java.time.LocalDate;
import java.time.Month;
import java.time.Year;
import java.time.YearMonth;
import java.util.*;

public class DateHandler {
    // Currently, does not take into account rent periods that cross between years (like dec 2022 to jan 2023)
    public static int getDaysOfMonth(int monthInt){
        int year = Year.now().getValue();
        Month month = Month.of(monthInt);
        YearMonth yearMonth = YearMonth.of(year, month);
        return yearMonth.lengthOfMonth();
    }

    public static int getDaysOfMonth(int yearInt, int monthInt){
        Month month = Month.of(monthInt);
        YearMonth yearMonth = YearMonth.of(yearInt, month);
        return yearMonth.lengthOfMonth();
    }

    // Generated a map which links the name of the months with their numeral version
    public static Map<String, Integer> generateMonthsMapper(){
        Map<String ,Integer> monthsMap = new LinkedHashMap<String, Integer>();
        monthsMap.put("January",    Calendar.JANUARY+1);
        monthsMap.put("February",   Calendar.FEBRUARY+1);
        monthsMap.put("March",      Calendar.MARCH+1);
        monthsMap.put("April",      Calendar.APRIL+1);
        monthsMap.put("May",        Calendar.MAY+1);
        monthsMap.put("June",       Calendar.JUNE+1);
        monthsMap.put("July",       Calendar.JULY+1);
        monthsMap.put("August",     Calendar.AUGUST+1);
        monthsMap.put("September",  Calendar.SEPTEMBER+1);
        monthsMap.put("October"  ,  Calendar.OCTOBER+1);
        monthsMap.put("November",   Calendar.NOVEMBER+1);
        monthsMap.put("December",   Calendar.DECEMBER+1);

        return monthsMap;
    }

    // Updates the days within the ComboBoxModel depending on the month selected
    public static void updateDaysInComboBox(Object o, DefaultComboBoxModel<Integer> defaultComboBoxModel){
        defaultComboBoxModel.removeAllElements();
        int monthInt = DateHandler.generateMonthsMapper().get((String)o);
        int daysInMonth = DateHandler.getDaysOfMonth(monthInt);
        for (int day = 1; day <= daysInMonth; day++){
            defaultComboBoxModel.addElement(day);
        }
    }

    // Checks if the given startDate and endDate are valid (e.g endDate is not before startDate)
    public static boolean dateIsValid(LocalDate startDate, LocalDate endDate){
        if (startDate.isBefore(endDate)) return true;
        else return startDate.isEqual(endDate);
    }

    // Checks if the given rent period is within the date of the given car's listing
    public static boolean rentPeriodIsValid(Car car, LocalDate startOfRentPeriod, LocalDate endOfRentPeriod){
        Listing listing = car.getListing();
        if(!listing.isAvailable()) return false;
        if(!dateIsValid(startOfRentPeriod, endOfRentPeriod)) return false;
        else return dateIsValid(listing.getStartDate(), startOfRentPeriod)
                && dateIsValid(endOfRentPeriod, listing.getEndDate());
    }
}
