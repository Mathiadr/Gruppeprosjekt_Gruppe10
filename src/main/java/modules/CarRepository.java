package modules;

import tools.CarRepositoryFileHandler;

import java.io.File;
import java.util.ArrayList;

public class CarRepository {
    private String repositoryName; // This holds the name of the repository being accessed.
    private ArrayList<Car> carArrayList = new ArrayList<>();
    private final CarRepositoryFileHandler carRepositoryFileHandler = new CarRepositoryFileHandler();

    public CarRepository(String repositoryName) {
        this.repositoryName = repositoryName;

        carArrayList = readFromJSON();
    }



    // Adds a new car to the repository file
    public void AddNewCar(Car car){
        //TODO: implement
        boolean exists = false;
        for (Car i : carArrayList){
            //Checks if the car being added does not already exist within the list
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                exists = true;
                break;
            }
        }
        if (!exists){
            //System.out.println("Adding car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " to carArrayList");
            this.carArrayList.add(car);
        }
        else{
            //System.out.println("Car " + car.getRegistrationNumber() + " Already exists");
        }
    }

    // Removes Car object from carArrayList
    public void RemoveExistingCar(Car car){
        if (carArrayList.contains(car)){
            //System.out.println("Removed car " + car.getRegistrationNumber() + " with car owner " + car.getOwner() + " from carArrayList");
            carArrayList.remove(car);
        }
        else{
            // TODO: Implement error message to GUI
            //System.out.println("Could not find car " + car.getRegistrationNumber() + " in CarArrayList");
        }
    }

    // It rents da car, dumbass
    public void RentCar(Car car){
        // TODO: Test if works
        try{
            for (Car i : GetAllAvailableCars()){
                if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                    //TODO: ADD EXTRA MEASURE TO ENSURE IT DOES NOT RENT AN UNAVAILABLE CAR
                    i.getListing().setAvailable(false);
                }
            }
        }
        catch (RuntimeException couldNotRentCar){
            //System.out.println("\n\nCould not rent car " + car.getRegistrationNumber() + "\n\n");
        }

    }

    // Returns all cars that are not currently occupied
    public ArrayList<Car> GetAllAvailableCars(){
        ArrayList<Car> allAvailableCars = new ArrayList<>();
        for (Car i : carArrayList){
            if (i.getListing().isAvailable()){
                allAvailableCars.add(i);
            }
        }
        return allAvailableCars;
    }

    public void updateListing(Car car, Listing listing){
        for (Car i : carArrayList){
            if (car.getRegistrationNumber().equals(i.getRegistrationNumber())){
                i.setListing(listing);
                //System.out.println("Updated listing in Car " + i.getRegistrationNumber());
                break;
            }
        }
    }

    public void SaveCarsToJSON(){
        carRepositoryFileHandler.writeArrayListToFile(getCarArrayList(), repositoryName);
    }

    // Reads all objects from JSON file and maps it to the Car object
    public ArrayList<Car> readFromJSON(){
        return carRepositoryFileHandler.readFromFile(repositoryName);
    }

    public String getRepositoryName() {
        return repositoryName;
    }

    public void setRepositoryName(String repositoryName) {
        this.repositoryName = repositoryName;
    }

    public ArrayList<Car> getCarArrayList() {
        return carArrayList;
    }

    public void setCarArrayList(ArrayList<Car> carArrayList) {
        this.carArrayList = carArrayList;
    }
}
