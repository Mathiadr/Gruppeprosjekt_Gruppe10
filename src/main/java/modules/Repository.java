package modules;

import tools.RepositoryFileHandler;

import java.util.ArrayList;
import java.util.Objects;

public class Repository {
    private String repositoryName; // This holds the name of the repository being accessed.
    private ArrayList<Listing> listingArrayList = new ArrayList<Listing>();

    private final RepositoryFileHandler repositoryFileHandler = new RepositoryFileHandler();

    public Repository(String repositoryName) {
        this.repositoryName = repositoryName;

        listingArrayList = readFromJSON();
    }



    // Reads all objects from JSON file and returns an ArrayList of Listing objects
    public ArrayList<Listing> readFromJSON(){
        return repositoryFileHandler.readFromFile(repositoryName);
    }


    public void addNewListing(Listing listing){
        boolean exists = false;
        for (Listing i : listingArrayList){
            //TODO: IMPLEMENT
            if(Objects.equals(i.getCar().getRegistrationNumber(), listing.getCar().getRegistrationNumber())){
                exists = true;
                break;
            }
        }
        if(!exists){
            listingArrayList.add(listing);
        }

    }

    // Removes Car object from listingArrayList
    public void removeExistingListing(Car car){
        // TODO: IMPLEMENT
    }

    // It rents da car, dumbass
    public void acceptListing(Car car){
        //TODO: REMOVE & RENAME
    }

    // Returns all cars that are not currently occupied
    public ArrayList<Listing> GetAllAvailableListings(){
        // TODO: Add date as parameter argument
        ArrayList<Listing> allAvailableListings = new ArrayList<>();
        for (Listing i : listingArrayList){
            if (i.isAvailable()){
                allAvailableListings.add(i);
            }
        }
        return allAvailableListings;
    }

    public void SaveCarsToJSON(){
        repositoryFileHandler.writeArrayListToFile(getListingArrayList(), repositoryName);
    }

    public String getRepositoryName() {
        return repositoryName;
    }

    public void setRepositoryName(String repositoryName) {
        this.repositoryName = repositoryName;
    }

    public ArrayList<Listing> getListingArrayList() {
        return listingArrayList;
    }

    public void setListingArrayList(ArrayList<Listing> listingArrayList) {
        this.listingArrayList = listingArrayList;
    }
}
