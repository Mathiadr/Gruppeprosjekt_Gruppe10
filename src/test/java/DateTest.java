import modules.Car;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;
import tools.DateHandler;
import java.time.LocalDate;
import java.time.Month;
import java.time.Year;
import java.time.YearMonth;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class DateTest {

    // Provides an array of dates to compare with
    public static Stream<Arguments> dateSet(){
        return Stream.of(
                Arguments.of((Object[]) new LocalDate[][]{{
                        LocalDate.of(2022, 1, 1),
                        LocalDate.of(2022, 12, 31),
                        LocalDate.of(2022, 6, 1),
                        LocalDate.of(2022, 7, 3)
                }}),
                Arguments.of((Object[]) new LocalDate[][]{{
                        LocalDate.of(2020, 2, 2),
                        LocalDate.of(2021, 2, 2),
                        LocalDate.of(2020, 2, 2),
                        LocalDate.of(2021, 2, 2)
                }}),
                Arguments.of((Object[]) new LocalDate[][]{{
                        LocalDate.of(2019, 6, 15),
                        LocalDate.of(2019, 12, 25),
                        LocalDate.of(2019, 8, 15),
                        LocalDate.of(2019, 9, 25),
                }}),
                Arguments.of((Object[]) new LocalDate[][]{{
                        LocalDate.of(2020, 11, 11),
                        LocalDate.of(2025, 11, 11),
                        LocalDate.of(2024, 1, 1),
                        LocalDate.of(2024, 1, 31),
                }}));
    }

    @ParameterizedTest
    @MethodSource(value = "dateSet")
    public void chronological_date_comparison_is_valid(LocalDate[] localDates){
        int startDate = 0;
        int endDate = 1;
        assertTrue(DateHandler.dateIsValid(localDates[startDate], localDates[endDate]));
    }

    @ParameterizedTest
    @MethodSource(value = "dateSet")
    public void non_chronological_date_comparison_is_invalid(LocalDate[] localDates){
        int startDate = 0;
        int endDate = 1;
        assertFalse(DateHandler.dateIsValid(localDates[endDate], localDates[startDate]));
    }

    @ParameterizedTest
    @MethodSource(value = "dateSet")
    public void same_date_comparison_is_valid(LocalDate[] localDates){
        int date = 0;
        assertTrue(DateHandler.dateIsValid(localDates[date], localDates[date]));
    }


    @Test
    public void month_number_to_string_mapper_is_accurate(){
        Map<String, Integer> monthMap = DateHandler.generateMonthsMapper();
        Approvals.verify(monthMap);
    }

    @Test
    public void month_has_correct_amount_of_days_with_2022_as_year(){
        int year = 2022;
        String[] months = {"January", "February", "March", "April", "May", "June",
                "July", "August", "September", "October"  , "November", "December"};
        Map<String, Integer> amountOfDaysPerMonthMap = new LinkedHashMap<>();
        for (int i = 0; i < 12; i++){
            amountOfDaysPerMonthMap.put(months[i], DateHandler.getDaysOfMonth(year, i+1));
        }
        Approvals.verify(amountOfDaysPerMonthMap);
    }

}
