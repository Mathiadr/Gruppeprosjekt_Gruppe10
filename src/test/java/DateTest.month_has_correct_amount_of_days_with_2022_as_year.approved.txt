April : 30 
August : 31 
December : 31 
February : 28 
January : 31 
July : 31 
June : 30 
March : 31 
May : 31 
November : 30 
October : 31 
September : 30 
